### PR TITLE
Hood Class - Periodic and Zero Hood

### PR DESCRIPTION
**Pull request by Aadit-Ambadkar**
Mostly Copied code, but the periodic is done, and the zerohood is done. To zero the hood, call setZero();

On periodic automatically displays hoodEncoder position to smart dashboard.

@jaredgoodman03 is there anything else I need to do? If not, then we need to figure out manual vs camera, which I need your help for.

### CODE REVIEW

**Comment by KTOmega** [comment]
Do we need some of these files?
- `2021-offseason/src/main/java/com/team254/lib/util/MA3AnalogEncoder.java`
- `2021-offseason/src/main/java/com/team254/lib/util/MA3Encoder.java`
- `2021-offseason/src/main/java/com/team254/lib/util/AdaptivePurePursuitController.java`
- `2021-offseason/src/main/java/com/team254/lib/util/RevRoboticsAirPressureSensor.java`

These probably should have been added in a separate PR.

File: 2021-offseason/src/main/java/frc/robot/commands/Hood.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        
+    }
+    @Override
+    public void periodic() {
+        SmartDashboard.putNumber("real output", hoodMain.getAppliedOutput());
+        SmartDashboard.putNumber("bruh position", hoodEncoder.getPosition());
+        if (isZeroOn) tareHood();
+    }
+
+    public void setZero() {
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/commands/Hood.java:57]: nit: naming: should be more descriptive, consider `startZeroHood` or `startTareHood`

File: 2021-offseason/src/main/java/frc/robot/commands/Hood.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    private void setSetpoint(double setpoint) {
+        this.hoodController.setReference(setpoint, ControlType.kPosition);
+    }
+
+    public void increaseSetpoint(double amount) {
+        curdeg+=amount;
+        curdeg=Math.max(0, curdeg);
+        setSetpoint(curdeg+Constants.Hood.INITIAL_SETPOINT);
+    }
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/commands/Hood.java:78]: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/commands/Hood.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        this.hoodController.setI(Constants.Hood.kI);
+        this.hoodController.setD(Constants.Hood.kD);
+        this.camera = new PhotonCamera("myCamera");
+        //this.setSetpoint(50);
+        
+    }
+    @Override
+    public void periodic() {
+        SmartDashboard.putNumber("real output", hoodMain.getAppliedOutput());
+        SmartDashboard.putNumber("bruh position", hoodEncoder.getPosition());
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/commands/Hood.java:53]: I'd prefer it if we removed these, but I can see how these are useful. if you decide to keep these, please name them better, like `hood applied output` and `hood encoder position`

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -4,12 +4,54 @@
...
  *
- * <p>It is advised to statically import this class (or one of its inner classes) wherever the
- * constants are needed, to reduce verbosity.
+ * <p>
+ * It is advised to statically import this class (or one of its inner classes)
+ * wherever the constants are needed, to reduce verbosity.
  */
-public final class Constants {}
+public final class Constants {
+    public static final class Hood {
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:22]: `kHood`

**Review by KTOmega — CHANGES_REQUESTED** [review]
mostly style/naming changes

File: 2021-offseason/src/main/java/frc/robot/commands/Hood.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if (isZeroOn) tareHood();
+    }
+
+    public void setZero() {
+        isZeroOn = true;
+    }
+
+    private void tareHood() {
```
[the remark this answers is not in the document]
Comment (Aadit-Ambadkar) [reply 2021-offseason/src/main/java/frc/robot/commands/Hood.java:61]: We only use it once at the start of the match I believe, but I think the limit switch will be used.

File: 2021-offseason/src/main/java/frc/robot/commands/Hood.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+public class Hood extends SubsystemBase {
+    private CANSparkMax hoodMain;
+    private CANEncoder hoodEncoder;
+    private CANPIDController hoodController;
+    private InterpolatingTreeMap closeTreeMap;
+    private InterpolatingTreeMap farTreeMap;
+    private PhotonCamera camera;
+    private double curdeg = 0;
+    private DigitalInput limitswitch = new DigitalInput(Constants.Hood.LIMIT_PORT);
+    private boolean isZeroOn = false;
```
[the remark this answers is not in the document]
Comment (Aadit-Ambadkar) [reply 2021-offseason/src/main/java/frc/robot/commands/Hood.java:33]: hoodMain, etc are defined in the constructor though

[the remark this answers is not in the document]
Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/commands/Hood.java:33]: `final` variables can be set in the constructor

**Comment by KTOmega** [comment]
can you batch your commits instead of deleting one by one lol

**Review by jaredgoodman03 — APPROVED** [review]

**Review by jaredgoodman03 — APPROVED** [review]

**Review by jaredgoodman03 — APPROVED** [review]

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        initialSetpoint = hoodEncoder.getPosition();
+    }
+    @Override
+    public void periodic() {
+        SmartDashboard.putNumber("hood applied output", hoodMain.getAppliedOutput());
+        SmartDashboard.putNumber("hood encoder position", hoodEncoder.getPosition());
+        if (isTaring) tareHoodPeriodic();
+    }
+
+    public void setZero() {
```
Comment (poroheart) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:59]: you should call this method in the constructor to zero the hood at the start of the match

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    public void incrementDown() {
+        increaseSetpoint(-Constants.kHood.HOOD_INCREMENT);
+    }
+    public void increaseSetpoint(double amount) {
+        curdeg+=amount;
+        curdeg=Math.max(0, curdeg);
+        set(curdeg+initialSetpoint);
+    }
+    public void setSetpoint(double setpoint) {
+        increaseSetpoint(setpoint-curdeg);
```
Comment (poroheart) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:88]: for this method just use the set() method

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.photonvision.PhotonCamera;
+
+public class Hood extends SubsystemBase {
+    private final CANSparkMax hoodMain;
+    private final CANEncoder hoodEncoder;
+    private final CANPIDController hoodController;
+    private InterpolatingTreeMap closeTreeMap;
+    private InterpolatingTreeMap farTreeMap;
+    private PhotonCamera camera;
+    private double curdeg = 0;
```
Comment (poroheart) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:33]: at least camel case this variable name so its easy to tell what it is just from the name

**Review by poroheart — CHANGES_REQUESTED** [review]

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java (hunk 8 above)
Comment (Aadit-Ambadkar) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:88]: I used increase so it doesn't go too far down

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java (hunk 9 above)
Comment (Aadit-Ambadkar) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:33]: which variable?

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java (hunk 8 above)
Comment (poroheart) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:89]: why would it go too far down

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+public class Hood extends SubsystemBase {
+    private final CANSparkMax hoodMain;
+    private final CANEncoder hoodEncoder;
+    private final CANPIDController hoodController;
+    private InterpolatingTreeMap closeTreeMap;
+    private InterpolatingTreeMap farTreeMap;
+    private PhotonCamera camera;
+    private double currentDegree = 0;
+    private final DigitalInput limitswitch = new DigitalInput(Constants.kHood.LIMIT_PORT);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:34]: nit: camel case: `limitSwitch`

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        this.hoodEncoder = this.hoodMain.getEncoder();
+        this.hoodController = this.hoodMain.getPIDController();
+        this.hoodController.setOutputRange(-Constants.kHood.MAX_SPEED, Constants.kHood.MAX_SPEED);
+        this.hoodController.setP(Constants.kHood.kP);
+        this.hoodController.setI(Constants.kHood.kI);
+        this.hoodController.setD(Constants.kHood.kD);
+        this.camera = new PhotonCamera("myCamera");
+        initialSetpoint = hoodEncoder.getPosition();
+        setZero();
+    }
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:52]: nit: spacing: newline after end of method/ctor decl

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    private void tareHoodPeriodic() {
+        if (limitswitch.get()) {
+            hoodMain.set(0);
+            currentDegree = 0;
+            isTaring = false;
+            initialSetpoint = hoodEncoder.getPosition();
+        }
+        else hoodMain.set(Constants.kHood.MAX_SPEED);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:71]: nit: styling
```suggestion
        } else {
            hoodMain.set(Constants.kHood.MAX_SPEED);
        }
```

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public void set(double setpoint) {
+        this.hoodController.setReference(setpoint, ControlType.kPosition);
+    }
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:76]: nit: spacing: newline after end of method/ctor decl
same for the rest of these methods

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    public void incrementUp() {
+        increaseSetpoint(Constants.kHood.HOOD_INCREMENT);
+    }
+    public void incrementDown() {
+        increaseSetpoint(-Constants.kHood.HOOD_INCREMENT);
+    }
+    public void increaseSetpoint(double amount) {
+        currentDegree+=amount;
+        currentDegree=Math.max(0, currentDegree);
+        set(currentDegree+initialSetpoint);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:86]: nit: spacing between operators
```suggestion
        currentDegree += amount;
        currentDegree = Math.max(0, currentDegree);
        set(currentDegree + initialSetpoint);
```

Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:86]: also line 89

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        SmartDashboard.putNumber("kD", Constants.kHood.kD);
+        this.hoodMain = new CANSparkMax(Constants.kHood.MOTOR_ID, Constants.kHood.MOTOR_TYPE);
+        hoodMain.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        this.hoodEncoder = this.hoodMain.getEncoder();
+        this.hoodController = this.hoodMain.getPIDController();
+        this.hoodController.setOutputRange(-Constants.kHood.MAX_SPEED, Constants.kHood.MAX_SPEED);
+        this.hoodController.setP(Constants.kHood.kP);
+        this.hoodController.setI(Constants.kHood.kI);
+        this.hoodController.setD(Constants.kHood.kD);
+        this.camera = new PhotonCamera("myCamera");
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:49]: consider making `"myCamera"` a constant

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import org.photonvision.PhotonCamera;
+
+public class Hood extends SubsystemBase {
+    private final CANSparkMax hoodMain;
+    private final CANEncoder hoodEncoder;
+    private final CANPIDController hoodController;
+    private InterpolatingTreeMap closeTreeMap;
+    private InterpolatingTreeMap farTreeMap;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:31]: variables declared but never assigned to nor used

File: 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    private PhotonCamera camera;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Hood.java:32]: nit: make this `final`

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 18]
```diff
@@ -15,6 +18,43 @@
...
+        public static final CANSparkMaxLowLevel.MotorType MOTOR_TYPE = CANSparkMaxLowLevel.MotorType.kBrushless;
+        public static final double MAX_SPEED = 0.3;
+        public static final double HOOD_INCREMENT = 10;
+        public static final double kP = 0.1;
+        public static final double kI = 0;
+        public static final double kD = 0;
+        public static final double SETPOINT = (45.0/360.0) * (332.0/14.0) * 15;
+        public static final InterpolatingTreeMap<InterpolatingDouble, InterpolatingDouble> hoodAngleTreeMap = new InterpolatingTreeMap<InterpolatingDouble, InterpolatingDouble>();
+        public static final double[] ZONE_SETPOINTS = {37.0, 56.0, 80.0, 80.0};
+        public static final int LIMIT_PORT = -1;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:33]: nit: naming: consider `LIMIT_SWITCH_PORT`

**Review by KTOmega — COMMENTED** [review]
a few nitpicks

**Pull request merged**